src/internal/_toString.js

```diff
@@ -0,0 +1,36 @@
+var _indexOf = require('./_indexOf');
+var _map = require('./_map');
+var _quote = require('./_quote');
+var _toISOString = require('./_toISOString');
+var keys = require('../keys');
+
+
+module.exports = function _toString(x, seen) {
+    var recur = function recur(y) {
+        var xs = seen.concat([x]);
+        return _indexOf(xs, y) >= 0 ? '<Circular>' : _toString(y, xs);
+    };
+
+    switch (Object.prototype.toString.call(x)) {
+        case '[object Arguments]':
+            return '(function() { return arguments; }(' + _map(recur, x).join(', ') + '))';
+        case '[object Array]':
+            return '[' + _map(recur, x).join(', ') + ']';
+        case '[object Boolean]':
+            return typeof x === 'object' ? 'new Boolean(' + recur(x.valueOf()) + ')' : x.toString();
+        case '[object Date]':
+            return 'new Date(' + _quote(_toISOString(x)) + ')';
+        case '[object Null]':
+            return 'null';
+        case '[object Number]':
+            return typeof x === 'object' ? 'new Number(' + recur(x.valueOf()) + ')' : 1 / x === -Infinity ? '-0' : x.toString(10);
+        case '[object String]':
+            return typeof x === 'object' ? 'new String(' + recur(x.valueOf()) + ')' : _quote(x);
+        case '[object Undefined]':
+            return 'undefined';
+        default:
+            return x.toString !== Object.prototype.toString ?  // Boolean, Function, RegExp, user-defined types
```

Comment:
I think this should be really added as a ramda extension as it doesn't have
a clear purpose for general programs and its debugging functionality is
mostly covered by console.
 On Mar 20, 2015 4:41 PM, "David Chambers" <notifications@github.com> wrote:

> In src/internal/_toString.js
> <https://github.com/ramda/ramda/pull/924#discussion_r26874458>:
>
> > +        case '[object Array]':
> > +            return '[' + _map(recur, x).join(', ') + ']';
> > +        case '[object Boolean]':
> > +            return typeof x === 'object' ? 'new Boolean(' + recur(x.valueOf()) + ')' : x.toString();
> > +        case '[object Date]':
> > +            return 'new Date(' + _quote(_toISOString(x)) + ')';
> > +        case '[object Null]':
> > +            return 'null';
> > +        case '[object Number]':
> > +            return typeof x === 'object' ? 'new Number(' + recur(x.valueOf()) + ')' : 1 / x === -Infinity ? '-0' : x.toString(10);
> > +        case '[object String]':
> > +            return typeof x === 'object' ? 'new String(' + recur(x.valueOf()) + ')' : _quote(x);
> > +        case '[object Undefined]':
> > +            return 'undefined';
> > +        default:
> > +            return x.toString !== Object.prototype.toString ?  // Boolean, Function, RegExp, user-defined types
>
> Good idea.
>
> x.toString() will throw if x is an object created via Object.create(null),
> so we first need to check whether typeof x.toString === 'function'. I
> added a test case for this.
>
> As well as checking that x.toString() !== '[object Object]' we can check
> that x.constructor.name !== 'Object'. I've updated the pull request to
> include both checks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ramda/ramda/pull/924/files#r26874458>.
>